README.md
## RcppFaddeeva [![Build Status](https://travis-ci.org/baptiste/rcppfaddeeva.svg)](https://travis-ci.org/baptiste/rcppfaddeeva)

[![](http://www.r-pkg.org/badges/version/RcppFaddeeva)](http://cran.rstudio.com/web/packages/RcppFaddeeva/index.html)
[![](http://cranlogs.r-pkg.org/badges/grand-total/RcppFaddeeva)](http://cran.rstudio.com/web/packages/RcppFaddeeva/index.html)

Wrapper to the
[Faddeeva Package by Steven G. Johnson](http://ab-initio.mit.edu/wiki/index.php/Faddeeva_Package)

## What is Faddeeva?

The [Faddeeva](http://ab-initio.mit.edu/wiki/index.php/Faddeeva_Package)
package provides C++ code to compute the various error functions of arbitrary
complex arguments, in particular the Faddeeva function, error function,
complementary error function, scaled complementary error function and
imaginary error function as well as the Dawson function.  With these, Voigt
functions, Fresnel integrals and more can be computed.

### Why this package?

The package was created to provide a fast implementation of the Voigt lineshape, formally defined as the convolution between a Lorentzian and a Gaussian. The imaginary part of the Faddeeva function, properly scaled, provides directly the result.

This package also results in a nice example for creating Rcpp wrappers for a small and
self-contained C++ library, in this case a single header file along with a
single code file.

## Authors

Baptiste Auguié and Dirk Eddelbuettel

### License

[![License](http://img.shields.io/badge/license-GPL%20%28%3E=%202%29-brightgreen.svg?style=flat)](http://www.gnu.org/licenses/gpl-2.0.html)


inst/NEWS.md

# RcppFaddeeva 0.2.1 (09/2016)

INTERNAL CHANGES

* tests with testthat moved to correct location

# RcppFaddeeva 0.2.0 (13/06/2016)

* C++ API, contributed by Dennis Feehan

# RcppFaddeeva 0.1.1 (1/07/2015)

* Switch vignette engine to `rmarkdown`
* explicitly import `dnorm` from `stats`

# RcppFaddeeva 0.1.0 (9/06/2015)

* Initial version


tests/minimal.R
library(RcppFaddeeva)
# ?RcppFaddeeva

z <- 1:10 + 1i

erfcx(z)
erf(z)
erfi(z)
erfc(z)
Dawson(z)
Faddeeva_w(z)


tests/test-voigt.R
library(RcppFaddeeva)

## should integrate to 1
integrate(Voigt, -Inf, Inf, x0=400, sigma=50, gamma=100)

x <- seq(-1000, 1000)
x0 <- 200
v <- Voigt(x, x0, 100, 30)
l <- Lorentz(x, x0, 30)
g <- Gauss(x, x0, 100)
N <- length(x)
c <- convolve(Gauss(x, 200, 100), Lorentz(x, 0, 30), type="o")[seq(N/2, length=N)]

matplot(x, cbind(v, l, g, c), t="l", lty=c(1,2,2,1), xlab="x", ylab="")
legend("topleft", legend = c("Voigt", "Lorentz", "Gauss", "convolution"), bty="n",
       lty=c(1,2,2,1), col=1:4)


inst/tests/test-voigt.R
library(RcppFaddeeva)

## should integrate to 1
integrate(Voigt, -Inf, Inf, x0=400, sigma=50, gamma=100)

x <- seq(-1000, 1000)
x0 <- 200
v <- Voigt(x, x0, 100, 30)
l <- Lorentz(x, x0, 30)
g <- Gauss(x, x0, 100)
N <- length(x)
c <- convolve(RcppFaddeeva:::Gauss(x, 200, 100), 
              RcppFaddeeva:::Lorentz(x, 0, 30), type="o")[seq(N/2, length=N)]

matplot(x, cbind(v, l, g, c), t="l", lty=c(1,2,2,1), xlab="x", ylab="")
legend("topleft", legend = c("Voigt", "Lorentz", "Gauss", "convolution"), bty="n",
       lty=c(1,2,2,1), col=1:4)


R/RcppExports.R
# This file was generated by Rcpp::compileAttributes
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' @title Faddeeva family of error functions of the complex variable
#' @description the Faddeeva function
#' @param z complex vector
#' @param relerr double, requested error
#' @return complex vector
#' @describeIn wrap compute w(z) = exp(-z^2) erfc(-iz)
#' @family wrapper
#' @examples
#' Faddeeva_w(1:10 + 1i)
#' @export
Faddeeva_w <- function(z, relerr = 0) {
    .Call('RcppFaddeeva_Faddeeva_w', PACKAGE = 'RcppFaddeeva', z, relerr)
}

#' the scaled complementary error function
#' @inheritParams Faddeeva_w
#' @describeIn wrap compute erfcx(z) = exp(z^2) erfc(z)
#' @family wrapper
#' @examples
#' erfcx(1:10 + 1i)
#' @export
erfcx <- function(z, relerr = 0) {
    .Call('RcppFaddeeva_erfcx', PACKAGE = 'RcppFaddeeva', z, relerr)
}

#'  the error function of complex arguments
#' @inheritParams Faddeeva_w
#' @describeIn wrap compute erf(z)
#' @family wrapper
#' @examples
#' erf(1:10 + 1i)
#' @export
erf <- function(z, relerr = 0) {
    .Call('RcppFaddeeva_erf', PACKAGE = 'RcppFaddeeva', z, relerr)
}

#' the imaginary error function
#' @inheritParams Faddeeva_w
#' @describeIn wrap compute erfi(z) = -i erf(iz)
#' @family wrapper
#' @examples
#' erfi(1:10 + 1i)
#' @export
erfi <- function(z, relerr = 0) {
    .Call('RcppFaddeeva_erfi', PACKAGE = 'RcppFaddeeva', z, relerr)
}

#' the complementary error function
#' @inheritParams Faddeeva_w
#' @describeIn wrap compute erfc(z) = 1 - erf(z)
#' @family wrapper
#' @examples
#' erfc(1:10 + 1i)
#' @export
erfc <- function(z, relerr = 0) {
    .Call('RcppFaddeeva_erfc', PACKAGE = 'RcppFaddeeva', z, relerr)
}

#' the Dawson function
#' @inheritParams Faddeeva_w
#' @describeIn wrap compute Dawson(z) = sqrt(pi)/2  *  exp(-z^2) * erfi(z)
#' @family wrapper
#' @examples
#' Dawson(1:10 + 1i)
#' @export
Dawson <- function(z, relerr = 0) {
    .Call('RcppFaddeeva_Dawson', PACKAGE = 'RcppFaddeeva', z, relerr)
}

# Register entry points for exported C++ functions
methods::setLoadAction(function(ns) {
    .Call('RcppFaddeeva_RcppExport_registerCCallable', PACKAGE = 'RcppFaddeeva')
})


src/RcppExports.cpp
// This file was generated by Rcpp::compileAttributes
// Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#include "../inst/include/RcppFaddeeva.h"
#include <Rcpp.h>
#include <string>
#include <set>

using namespace Rcpp;

// Faddeeva_w
std::vector< std::complex<double> > Faddeeva_w(const std::vector< std::complex<double> >& z, double relerr);
static SEXP RcppFaddeeva_Faddeeva_w_try(SEXP zSEXP, SEXP relerrSEXP) {
BEGIN_RCPP
    Rcpp::RObject __result;
    Rcpp::traits::input_parameter< const std::vector< std::complex<double> >& >::type z(zSEXP);
    Rcpp::traits::input_parameter< double >::type relerr(relerrSEXP);
    __result = Rcpp::wrap(Faddeeva_w(z, relerr));
    return __result;
END_RCPP_RETURN_ERROR
}
RcppExport SEXP RcppFaddeeva_Faddeeva_w(SEXP zSEXP, SEXP relerrSEXP) {
    SEXP __result;
    {
        Rcpp::RNGScope __rngScope;
        __result = PROTECT(RcppFaddeeva_Faddeeva_w_try(zSEXP, relerrSEXP));
    }
    Rboolean __isInterrupt = Rf_inherits(__result, "interrupted-error");
    if (__isInterrupt) {
        UNPROTECT(1);
        Rf_onintr();
    }
    Rboolean __isError = Rf_inherits(__result, "try-error");
    if (__isError) {
        SEXP __msgSEXP = Rf_asChar(__result);
        UNPROTECT(1);
        Rf_error(CHAR(__msgSEXP));
    }
    UNPROTECT(1);
    return __result;
}
// erfcx
std::vector< std::complex<double> > erfcx(const std::vector< std::complex<double> >& z, double relerr);
static SEXP RcppFaddeeva_erfcx_try(SEXP zSEXP, SEXP relerrSEXP) {
BEGIN_RCPP
    Rcpp::RObject __result;
    Rcpp::traits::input_parameter< const std::vector< std::complex<double> >& >::type z(zSEXP);
    Rcpp::traits::input_parameter< double >::type relerr(relerrSEXP);
    __result = Rcpp::wrap(erfcx(z, relerr));
    return __result;
END_RCPP_RETURN_ERROR
}
RcppExport SEXP RcppFaddeeva_erfcx(SEXP zSEXP, SEXP relerrSEXP) {
    SEXP __result;
    {
        Rcpp::RNGScope __rngScope;
        __result = PROTECT(RcppFaddeeva_erfcx_try(zSEXP, relerrSEXP));
    }
    Rboolean __isInterrupt = Rf_inherits(__result, "interrupted-error");
    if (__isInterrupt) {
        UNPROTECT(1);
        Rf_onintr();
    }
    Rboolean __isError = Rf_inherits(__result, "try-error");
    if (__isError) {
        SEXP __msgSEXP = Rf_asChar(__result);
        UNPROTECT(1);
        Rf_error(CHAR(__msgSEXP));
    }
    UNPROTECT(1);
    return __result;
}
// erf
std::vector< std::complex<double> > erf(const std::vector< std::complex<double> >& z, double relerr);
static SEXP RcppFaddeeva_erf_try(SEXP zSEXP, SEXP relerrSEXP) {
BEGIN_RCPP
    Rcpp::RObject __result;
    Rcpp::traits::input_parameter< const std::vector< std::complex<double> >& >::type z(zSEXP);
    Rcpp::traits::input_parameter< double >::type relerr(relerrSEXP);
    __result = Rcpp::wrap(erf(z, relerr));
    return __result;
END_RCPP_RETURN_ERROR
}
RcppExport SEXP RcppFaddeeva_erf(SEXP zSEXP, SEXP relerrSEXP) {
    SEXP __result;
    {
        Rcpp::RNGScope __rngScope;
        __result = PROTECT(RcppFaddeeva_erf_try(zSEXP, relerrSEXP));
    }
    Rboolean __isInterrupt = Rf_inherits(__result, "interrupted-error");
    if (__isInterrupt) {
        UNPROTECT(1);
        Rf_onintr();
    }
    Rboolean __isError = Rf_inherits(__result, "try-error");
    if (__isError) {
        SEXP __msgSEXP = Rf_asChar(__result);
        UNPROTECT(1);
        Rf_error(CHAR(__msgSEXP));
    }
    UNPROTECT(1);
    return __result;
}
// erfi
std::vector< std::complex<double> > erfi(const std::vector< std::complex<double> >& z, double relerr);
static SEXP RcppFaddeeva_erfi_try(SEXP zSEXP, SEXP relerrSEXP) {
BEGIN_RCPP
    Rcpp::RObject __result;
    Rcpp::traits::input_parameter< const std::vector< std::complex<double> >& >::type z(zSEXP);
    Rcpp::traits::input_parameter< double >::type relerr(relerrSEXP);
    __result = Rcpp::wrap(erfi(z, relerr));
    return __result;
END_RCPP_RETURN_ERROR
}
RcppExport SEXP RcppFaddeeva_erfi(SEXP zSEXP, SEXP relerrSEXP) {
    SEXP __result;
    {
        Rcpp::RNGScope __rngScope;
        __result = PROTECT(RcppFaddeeva_erfi_try(zSEXP, relerrSEXP));
    }
    Rboolean __isInterrupt = Rf_inherits(__result, "interrupted-error");
    if (__isInterrupt) {
        UNPROTECT(1);
        Rf_onintr();
    }
    Rboolean __isError = Rf_inherits(__result, "try-error");
    if (__isError) {
        SEXP __msgSEXP = Rf_asChar(__result);
        UNPROTECT(1);
        Rf_error(CHAR(__msgSEXP));
    }
    UNPROTECT(1);
    return __result;
}
// erfc
std::vector< std::complex<double> > erfc(const std::vector< std::complex<double> >& z, double relerr);
static SEXP RcppFaddeeva_erfc_try(SEXP zSEXP, SEXP relerrSEXP) {
BEGIN_RCPP
    Rcpp::RObject __result;
    Rcpp::traits::input_parameter< const std::vector< std::complex<double> >& >::type z(zSEXP);
    Rcpp::traits::input_parameter< double >::type relerr(relerrSEXP);
    __result = Rcpp::wrap(erfc(z, relerr));
    return __result;
END_RCPP_RETURN_ERROR
}
RcppExport SEXP RcppFaddeeva_erfc(SEXP zSEXP, SEXP relerrSEXP) {
    SEXP __result;
    {
        Rcpp::RNGScope __rngScope;
        __result = PROTECT(RcppFaddeeva_erfc_try(zSEXP, relerrSEXP));
    }
    Rboolean __isInterrupt = Rf_inherits(__result, "interrupted-error");
    if (__isInterrupt) {
        UNPROTECT(1);
        Rf_onintr();
    }
    Rboolean __isError = Rf_inherits(__result, "try-error");
    if (__isError) {
        SEXP __msgSEXP = Rf_asChar(__result);
        UNPROTECT(1);
        Rf_error(CHAR(__msgSEXP));
    }
    UNPROTECT(1);
    return __result;
}
// Dawson
std::vector< std::complex<double> > Dawson(const std::vector< std::complex<double> >& z, double relerr);
static SEXP RcppFaddeeva_Dawson_try(SEXP zSEXP, SEXP relerrSEXP) {
BEGIN_RCPP
    Rcpp::RObject __result;
    Rcpp::traits::input_parameter< const std::vector< std::complex<double> >& >::type z(zSEXP);
    Rcpp::traits::input_parameter< double >::type relerr(relerrSEXP);
    __result = Rcpp::wrap(Dawson(z, relerr));
    return __result;
END_RCPP_RETURN_ERROR
}
RcppExport SEXP RcppFaddeeva_Dawson(SEXP zSEXP, SEXP relerrSEXP) {
    SEXP __result;
    {
        Rcpp::RNGScope __rngScope;
        __result = PROTECT(RcppFaddeeva_Dawson_try(zSEXP, relerrSEXP));
    }
    Rboolean __isInterrupt = Rf_inherits(__result, "interrupted-error");
    if (__isInterrupt) {
        UNPROTECT(1);
        Rf_onintr();
    }
    Rboolean __isError = Rf_inherits(__result, "try-error");
    if (__isError) {
        SEXP __msgSEXP = Rf_asChar(__result);
        UNPROTECT(1);
        Rf_error(CHAR(__msgSEXP));
    }
    UNPROTECT(1);
    return __result;
}

// validate (ensure exported C++ functions exist before calling them)
static int RcppFaddeeva_RcppExport_validate(const char* sig) { 
    static std::set<std::string> signatures;
    if (signatures.empty()) {
        signatures.insert("std::vector< std::complex<double> >(*Faddeeva_w)(const std::vector< std::complex<double> >&,double)");
        signatures.insert("std::vector< std::complex<double> >(*erfcx)(const std::vector< std::complex<double> >&,double)");
        signatures.insert("std::vector< std::complex<double> >(*erf)(const std::vector< std::complex<double> >&,double)");
        signatures.insert("std::vector< std::complex<double> >(*erfi)(const std::vector< std::complex<double> >&,double)");
        signatures.insert("std::vector< std::complex<double> >(*erfc)(const std::vector< std::complex<double> >&,double)");
        signatures.insert("std::vector< std::complex<double> >(*Dawson)(const std::vector< std::complex<double> >&,double)");
    }
    return signatures.find(sig) != signatures.end();
}

// registerCCallable (register entry points for exported C++ functions)
RcppExport SEXP RcppFaddeeva_RcppExport_registerCCallable() { 
    R_RegisterCCallable("RcppFaddeeva", "RcppFaddeeva_Faddeeva_w", (DL_FUNC)RcppFaddeeva_Faddeeva_w_try);
    R_RegisterCCallable("RcppFaddeeva", "RcppFaddeeva_erfcx", (DL_FUNC)RcppFaddeeva_erfcx_try);
    R_RegisterCCallable("RcppFaddeeva", "RcppFaddeeva_erf", (DL_FUNC)RcppFaddeeva_erf_try);
    R_RegisterCCallable("RcppFaddeeva", "RcppFaddeeva_erfi", (DL_FUNC)RcppFaddeeva_erfi_try);
    R_RegisterCCallable("RcppFaddeeva", "RcppFaddeeva_erfc", (DL_FUNC)RcppFaddeeva_erfc_try);
    R_RegisterCCallable("RcppFaddeeva", "RcppFaddeeva_Dawson", (DL_FUNC)RcppFaddeeva_Dawson_try);
    R_RegisterCCallable("RcppFaddeeva", "RcppFaddeeva_RcppExport_validate", (DL_FUNC)RcppFaddeeva_RcppExport_validate);
    return R_NilValue;
}


src/callFaddeeva.cpp
#include <Rcpp.h>

using namespace Rcpp;

#include "Faddeeva.h"

//' @title Faddeeva family of error functions of the complex variable
//' @description the Faddeeva function
//' @param z complex vector
//' @param relerr double, requested error
//' @return complex vector
//' @describeIn wrap compute w(z) = exp(-z^2) erfc(-iz)
//' @family wrapper
//' @examples 
//' Faddeeva_w(1:10 + 1i)
//' @export
// [[Rcpp::export]]
ComplexVector Faddeeva_w(const ComplexVector& z, double relerr=0) { 
  std::vector< std::complex<double> > z_ = as< std::vector< std::complex<double> > >(z);
  int N = z.size(), i;
  std::vector< std::complex<double> > result(N);
  for(int i=0; i<N; i++) {
    result[i] = Faddeeva::w(z_[i], relerr);
  }
  return wrap(result);
}

//' the scaled complementary error function
//' @inheritParams Faddeeva_w 
//' @describeIn wrap compute erfcx(z) = exp(z^2) erfc(z)
//' @family wrapper
//' @examples 
//' erfcx(1:10 + 1i)
//' @export
// [[Rcpp::export]]
ComplexVector erfcx(const ComplexVector& z, double relerr=0) { 
  std::vector<std::complex<double> > z_ = as< std::vector< std::complex<double> > >(z);
  int N = z.size(), i;
  std::vector< std::complex<double> > result(N);
  for(int i=0; i<N; i++) {
    result[i] = Faddeeva::erfcx(z_[i], relerr);
  }
  return wrap(result);
}

//'  the error function of complex arguments
//' @inheritParams Faddeeva_w 
//' @describeIn wrap compute erf(z)
//' @family wrapper
//' @examples 
//' erf(1:10 + 1i)
//' @export
// [[Rcpp::export]]
ComplexVector erf(const ComplexVector& z, double relerr=0) { 
  std::vector<std::complex<double> > z_ = as< std::vector< std::complex<double> > >(z);
  int N = z.size(), i;
  std::vector< std::complex<double> > result(N);
  for(int i=0; i<N; i++) {
    result[i] = Faddeeva::erf(z_[i], relerr);
  }
  return wrap(result);
}

//' the imaginary error function 
//' @inheritParams Faddeeva_w 
//' @describeIn wrap compute erfi(z) = -i erf(iz)
//' @family wrapper
//' @examples 
//' erfi(1:10 + 1i)
//' @export
// [[Rcpp::export]]
ComplexVector erfi(const ComplexVector& z, double relerr=0) { 
  std::vector<std::complex<double> > z_ = as< std::vector< std::complex<double> > >(z);
  int N = z.size(), i;
  std::vector< std::complex<double> > result(N);
  for(int i=0; i<N; i++) {
    result[i] = Faddeeva::erfi(z_[i], relerr);
  }
  return wrap(result);
}

//' the complementary error function
//' @inheritParams Faddeeva_w 
//' @describeIn wrap compute erfc(z) = 1 - erf(z)
//' @family wrapper
//' @examples 
//' erfc(1:10 + 1i)
//' @export
// [[Rcpp::export]]
ComplexVector erfc(const ComplexVector& z, double relerr=0) { 
  std::vector< std::complex<double> > z_ = as< std::vector< std::complex<double> > >(z);
  int N = z.size(), i;
  std::vector< std::complex<double> > result(N);
  for(int i=0; i<N; i++) {
    result[i] = Faddeeva::erfc(z_[i], relerr);
  }
  return wrap(result);
}

//' the Dawson function
//' @inheritParams Faddeeva_w 
//' @describeIn wrap compute Dawson(z) = sqrt(pi)/2  *  exp(-z^2) * erfi(z)
//' @family wrapper
//' @examples 
//' Dawson(1:10 + 1i)
//' @export
// [[Rcpp::export]]
ComplexVector Dawson(const ComplexVector& z, double relerr=0) { 
  std::vector< std::complex<double> > z_ = as< std::vector< std::complex<double> > >(z);
  int N = z.size(), i;
  std::vector< std::complex<double> > result(N);
  for(int i=0; i<N; i++) {
    result[i] = Faddeeva::Dawson(z_[i], relerr);
  }
  return wrap(result);
}





inst/include/RcppFaddeeva_RcppExports.h
// This file was generated by Rcpp::compileAttributes
// Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#ifndef __RcppFaddeeva_RcppExports_h__
#define __RcppFaddeeva_RcppExports_h__

#include <Rcpp.h>

namespace RcppFaddeeva {

    using namespace Rcpp;

    namespace {
        void validateSignature(const char* sig) {
            Rcpp::Function require = Rcpp::Environment::base_env()["require"];
            require("RcppFaddeeva", Rcpp::Named("quietly") = true);
            typedef int(*Ptr_validate)(const char*);
            static Ptr_validate p_validate = (Ptr_validate)
                R_GetCCallable("RcppFaddeeva", "RcppFaddeeva_RcppExport_validate");
            if (!p_validate(sig)) {
                throw Rcpp::function_not_exported(
                    "C++ function with signature '" + std::string(sig) + "' not found in RcppFaddeeva");
            }
        }
    }

    inline std::vector< std::complex<double> > Faddeeva_w(const std::vector< std::complex<double> >& z, double relerr = 0) {
        typedef SEXP(*Ptr_Faddeeva_w)(SEXP,SEXP);
        static Ptr_Faddeeva_w p_Faddeeva_w = NULL;
        if (p_Faddeeva_w == NULL) {
            validateSignature("std::vector< std::complex<double> >(*Faddeeva_w)(const std::vector< std::complex<double> >&,double)");
            p_Faddeeva_w = (Ptr_Faddeeva_w)R_GetCCallable("RcppFaddeeva", "RcppFaddeeva_Faddeeva_w");
        }
        RObject __result;
        {
            RNGScope __rngScope;
            __result = p_Faddeeva_w(Rcpp::wrap(z), Rcpp::wrap(relerr));
        }
        if (__result.inherits("interrupted-error"))
            throw Rcpp::internal::InterruptedException();
        if (__result.inherits("try-error"))
            throw Rcpp::exception(as<std::string>(__result).c_str());
        return Rcpp::as<std::vector< std::complex<double> > >(__result);
    }

    inline std::vector< std::complex<double> > erfcx(const std::vector< std::complex<double> >& z, double relerr = 0) {
        typedef SEXP(*Ptr_erfcx)(SEXP,SEXP);
        static Ptr_erfcx p_erfcx = NULL;
        if (p_erfcx == NULL) {
            validateSignature("std::vector< std::complex<double> >(*erfcx)(const std::vector< std::complex<double> >&,double)");
            p_erfcx = (Ptr_erfcx)R_GetCCallable("RcppFaddeeva", "RcppFaddeeva_erfcx");
        }
        RObject __result;
        {
            RNGScope __rngScope;
            __result = p_erfcx(Rcpp::wrap(z), Rcpp::wrap(relerr));
        }
        if (__result.inherits("interrupted-error"))
            throw Rcpp::internal::InterruptedException();
        if (__result.inherits("try-error"))
            throw Rcpp::exception(as<std::string>(__result).c_str());
        return Rcpp::as<std::vector< std::complex<double> > >(__result);
    }

    inline std::vector< std::complex<double> > erf(const std::vector< std::complex<double> >& z, double relerr = 0) {
        typedef SEXP(*Ptr_erf)(SEXP,SEXP);
        static Ptr_erf p_erf = NULL;
        if (p_erf == NULL) {
            validateSignature("std::vector< std::complex<double> >(*erf)(const std::vector< std::complex<double> >&,double)");
            p_erf = (Ptr_erf)R_GetCCallable("RcppFaddeeva", "RcppFaddeeva_erf");
        }
        RObject __result;
        {
            RNGScope __rngScope;
            __result = p_erf(Rcpp::wrap(z), Rcpp::wrap(relerr));
        }
        if (__result.inherits("interrupted-error"))
            throw Rcpp::internal::InterruptedException();
        if (__result.inherits("try-error"))
            throw Rcpp::exception(as<std::string>(__result).c_str());
        return Rcpp::as<std::vector< std::complex<double> > >(__result);
    }

    inline std::vector< std::complex<double> > erfi(const std::vector< std::complex<double> >& z, double relerr = 0) {
        typedef SEXP(*Ptr_erfi)(SEXP,SEXP);
        static Ptr_erfi p_erfi = NULL;
        if (p_erfi == NULL) {
            validateSignature("std::vector< std::complex<double> >(*erfi)(const std::vector< std::complex<double> >&,double)");
            p_erfi = (Ptr_erfi)R_GetCCallable("RcppFaddeeva", "RcppFaddeeva_erfi");
        }
        RObject __result;
        {
            RNGScope __rngScope;
            __result = p_erfi(Rcpp::wrap(z), Rcpp::wrap(relerr));
        }
        if (__result.inherits("interrupted-error"))
            throw Rcpp::internal::InterruptedException();
        if (__result.inherits("try-error"))
            throw Rcpp::exception(as<std::string>(__result).c_str());
        return Rcpp::as<std::vector< std::complex<double> > >(__result);
    }

    inline std::vector< std::complex<double> > erfc(const std::vector< std::complex<double> >& z, double relerr = 0) {
        typedef SEXP(*Ptr_erfc)(SEXP,SEXP);
        static Ptr_erfc p_erfc = NULL;
        if (p_erfc == NULL) {
            validateSignature("std::vector< std::complex<double> >(*erfc)(const std::vector< std::complex<double> >&,double)");
            p_erfc = (Ptr_erfc)R_GetCCallable("RcppFaddeeva", "RcppFaddeeva_erfc");
        }
        RObject __result;
        {
            RNGScope __rngScope;
            __result = p_erfc(Rcpp::wrap(z), Rcpp::wrap(relerr));
        }
        if (__result.inherits("interrupted-error"))
            throw Rcpp::internal::InterruptedException();
        if (__result.inherits("try-error"))
            throw Rcpp::exception(as<std::string>(__result).c_str());
        return Rcpp::as<std::vector< std::complex<double> > >(__result);
    }

    inline std::vector< std::complex<double> > Dawson(const std::vector< std::complex<double> >& z, double relerr = 0) {
        typedef SEXP(*Ptr_Dawson)(SEXP,SEXP);
        static Ptr_Dawson p_Dawson = NULL;
        if (p_Dawson == NULL) {
            validateSignature("std::vector< std::complex<double> >(*Dawson)(const std::vector< std::complex<double> >&,double)");
            p_Dawson = (Ptr_Dawson)R_GetCCallable("RcppFaddeeva", "RcppFaddeeva_Dawson");
        }
        RObject __result;
        {
            RNGScope __rngScope;
            __result = p_Dawson(Rcpp::wrap(z), Rcpp::wrap(relerr));
        }
        if (__result.inherits("interrupted-error"))
            throw Rcpp::internal::InterruptedException();
        if (__result.inherits("try-error"))
            throw Rcpp::exception(as<std::string>(__result).c_str());
        return Rcpp::as<std::vector< std::complex<double> > >(__result);
    }

}

#endif // __RcppFaddeeva_RcppExports_h__


R/Voigt.R
##' @description Voigt distribution
##' @title The Voigt function, corresponding to the convolution of a lorentzian and a gaussian distribution
##' @param x numeric vector
##' @param x0 scalar, peak position
##' @param sigma parameter of the gaussian
##' @param gamma parameter of the lorentzian
##' @param real logical, return only the real part of the complex Faddeeva
##' @param ... passed to Faddeeva_w
##' @return numeric or complex vector
##' @author baptiste Auguie
##' @describeIn Voigt Voigt lineshape function
##' @family helper_function
##' @examples
##' ## should integrate to 1 in all cases
##' integrate(Voigt, -Inf, Inf, x0=400, sigma=50, gamma=100)
##' x <- seq(-1000, 1000)
##' x0 <- 200
##' l <- Lorentz(x, x0, 30)
##' g <- Gauss(x, x0, 100)
##' N <- length(x)
##' c <- convolve(Gauss(x, 200, 100), 
##'               Lorentz(x, 0, 30), type="o")[seq(N/2, length=N)]
##' v <- Voigt(x, x0, 100, 30)
##' matplot(x, cbind(v, l, g, c), t="l", lty=c(1,2,2,1), xlab="x", ylab="")
##' legend("topleft", legend = c("Voigt", "Lorentz", "Gauss", "Convolution"), bty="n",
##'       lty=c(1,2,2,1), col=1:4)
##' @export
Voigt <- function(x, x0, sigma, gamma, real = TRUE, ...){
  
  z <- (x - x0 + gamma*1i) / (sigma * sqrt(2))
  w <- Faddeeva_w(z, ...)
  if(real) return(Re(w) / (sigma * sqrt(2*pi))) else
    w / (sigma * sqrt(2*pi))
  
}

##' @description Lorentzian distribution
##' @title Lorentz
##' @describeIn Voigt Lorentzian lineshape function
##' @inheritParams Voigt 
##' @export
##' @family helper_function
Lorentz <- function(x, x0, gamma){
  gamma / (pi*((x -x0)^2 + gamma^2))
}

##' @description Gaussian distribution
##' @title Gauss 
##' @describeIn Voigt Gaussian lineshape function
##' @inheritParams Voigt
##' @export
##' @family helper_function
Gauss <- function(x, x0, sigma){
  dnorm(x, x0, sd = sigma)
}



